### PR TITLE
EJBs over HTTPS: Securing Jakarta Enterprise Beans with mutual TLS authentication

### PR DESCRIPTION
**Pull request by jessicarod7**

### CODE REVIEW

File: _posts/2022-06-15-ejb-over-https.adoc [hunk 1]
```diff
@@ -0,0 +1,340 @@
+---
+layout: post
+title: "EJBs over HTTPS: Securing Jakarta Enterprise Beans with mutual TLS authentication"
```
Comment (fjuma) [inline _posts/2022-06-15-ejb-over-https.adoc]: Just to prevent confusion with the EJB over HTTP feature, I think it would be good to remove "EJBs over HTTPS" from the title.

File: _posts/2022-06-15-ejb-over-https.adoc [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+date: 2022-06-15
+tags: ejb tls sasl https credential-store key-store trust-store
+synopsis: A demonstration on how to configure Elytron for secure connections to EJBs, using mutual TLS, SASL authentication mechanisms, and credential stores.
+author: cam-rod
+---
+
+The *https://github.com/wildfly-security-incubator/elytron-examples/tree/master/ejb-https[ejb-https]* example
```
Comment (fjuma) [inline _posts/2022-06-15-ejb-over-https.adoc]: Maybe we could update the directory name for the example to something like "ejb-mutual-tls"

File: _posts/2022-06-15-ejb-over-https.adoc [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+demonstrates how to use the Elytron subsystem to secure connections to EJBs with mutual TLS
```
Comment (fjuma) [inline _posts/2022-06-15-ejb-over-https.adoc]: s/how to use the Elytron subsystem to secure connections to EJBs/how to use Elytron to secure EJBs

File: _posts/2022-06-15-ejb-over-https.adoc [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+authentication. It also shows how to restrict usage of EJBs, implementing a FileSystem realm protected by SASL
```
Comment (fjuma) [inline _posts/2022-06-15-ejb-over-https.adoc]: s/how to restrict usage of EJBs/how to restrict access to EJBs

Comment (fjuma) [reply _posts/2022-06-15-ejb-over-https.adoc]: s/implementing a/configuring a

File: _posts/2022-06-15-ejb-over-https.adoc [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+https://wildfly-security.github.io/wildfly-elytron/blog/advanced-ejb-security/[configuring the FileSystem realm and SASL mechanism],
+and on https://wildfly-security.github.io/wildfly-elytron/blog/ejb-over-http/[connecting with HTTP].
+
+
+== An Overview of Mutual TLS authentication
+
+In a standard TLS handshake, the server presents its certificate to the client, which then inspects the contents and
+signature. If the certificate matches one that the client trusts, or is signed by an entity that it trusts, the server's
+identity is verified and a key exchange is performed to secure the session. In mutual TLS authentication, the client also
+presents its own certificate, which the server will then use to verify the identity of the client.
```
Comment (fjuma) [inline _posts/2022-06-15-ejb-over-https.adoc]: Nice to have this overview here!

File: _posts/2022-06-15-ejb-over-https.adoc [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+
+Using the Elytron subsystem, a WildFly server can form a mutual TLS connection with a remote client. To do so, the client
+will generate a certificate, which the server will import. The client will be configured to present the certificate when
+performing the TLS handshake, which the server will verify against its own copy.
```
Comment (fjuma) [inline _posts/2022-06-15-ejb-over-https.adoc]: s/verify against its own copy/verify using its truststore (or something along those lines)

File: _posts/2022-06-15-ejb-over-https.adoc [hunk 7]
```diff
@@ -0,0 +1,340 @@
...
+
+== Generating the Client Certificate
+
+In order for the client to use a certificate, it must be signed by a server-trusted entity. For the purposes of this
+demonstration, we will self-sign a certificate and then import it into the server's trust store. However, such
+certificates are generally treated as insecure since the identity cannot be externally verified. To use a certificate
+signed by a trusted Certificate Authority, take a look at this https://wildfly-security.github.io/wildfly-elytron/blog/obtaining-certificates-from-lets-encrypt-using-the-wildfly-cli/[blog post].
```
Comment (fjuma) [inline _posts/2022-06-15-ejb-over-https.adoc]: I like that you explained this.

File: _posts/2022-06-15-ejb-over-https.adoc [hunk 8]
```diff
@@ -0,0 +1,340 @@
...
+[source,shell]
+----
+$ keytool -genkeypair -alias client -keyalg RSA -keysize 2048 -validity 365 -keystore /PATH/TO/sslClient.keystore -dname "CN=client" -storepass clientKeySecret
+----
+
+Note that we have set the common name on the key to `client` (using the option `-dname "CN=client"`). The server will
+try to match the common name on the certificate to the hostname presented during the TLS handshake. If the two do not
+match, the connection will fail.
+
+Once the key pair is created, we use it to generate a self-signed certificate, and save it to a certificate file:
```
Comment (fjuma) [inline _posts/2022-06-15-ejb-over-https.adoc]: Just a small terminology thing, the -genkeypair operation actually generates the key pair and wraps the public key into an X509 self-signed certificate and stores it in the keystore. The exportcert operation just exports that certificate to a file.

File: _posts/2022-06-15-ejb-over-https.adoc [hunk 9]
```diff
@@ -0,0 +1,340 @@
...
+connect to it using the management CLI:
+
+[source,shell]
+----
+$ WILDFLY_HOME/bin/jboss-cli.sh --connect
+----
+
+(_Note: In this post, replace_  `WILDFLY_HOME` _with the actual path to your WildFly installation._)
+
+We will use the Elytron subsystem to create a key store to secure the server's key pair. The key store will be
```
Comment (fjuma) [inline _posts/2022-06-15-ejb-over-https.adoc]: s/secure/store

File: _posts/2022-06-15-ejb-over-https.adoc [hunk 10]
```diff
@@ -0,0 +1,340 @@
...
+a self-signed certificate, we will not validate its authenticity.
+
+[source]
+----
+/subsystem=elytron/key-store=sslTrustStore:import-certificate(alias=client,path=/PATH/TO/sslClient.cer,credential-reference={clear-text=serverTrustSecret},trust-cacerts=true,validate=false)
+
+/subsystem=elytron/key-store=sslTrustStore:store()
+----
+
+We also configure a trust manager, which will access the certificate during the TLS handshake:
```
Comment (fjuma) [inline _posts/2022-06-15-ejb-over-https.adoc]: s/access/verify

File: _posts/2022-06-15-ejb-over-https.adoc [hunk 11]
```diff
@@ -0,0 +1,340 @@
...
+example. We use the *keytool* once again, identify the server's certificate by its alias and importing it into a client
+trust store:
+
+[source,shell]
+----
+$ keytool -importcert -keystore /PATH/TO/sslClient.truststore -storepass clientTrustSecret -alias localhost -trustcacerts -file /WILDFLY_HOME/standalone/configuration/sslServer.cer -noprompt
+----
+
+At this point, the client and server both have their corresponding key pairs, and trust the certificates of the other.
+The TLS configuration is complete.
```
Comment (fjuma) [inline _posts/2022-06-15-ejb-over-https.adoc]: The explanations above for each step in the configuration are great!

File: _posts/2022-06-15-ejb-over-https.adoc [hunk 12]
```diff
@@ -0,0 +1,340 @@
...
+
+== Configuring the Credential Store
+
+We will now setup the client login using the SASL authentication mechanism. First, we use the *elytron-tool* generate a
```
Comment (fjuma) [inline _posts/2022-06-15-ejb-over-https.adoc]: s/generate/to generate

File: _posts/2022-06-15-ejb-over-https.adoc [hunk 13]
```diff
@@ -0,0 +1,340 @@
...
+credential store to securely save the login details:
```
Comment (fjuma) [inline _posts/2022-06-15-ejb-over-https.adoc]: s/save/store

File: _posts/2022-06-15-ejb-over-https.adoc [hunk 14]
```diff
@@ -0,0 +1,340 @@
...
+Then, we add the `example_user` identity to this credential store:
+
+[source,shell]
+----
+$ WILDFLY_HOME/bin/elytron-tool.sh credential-store --location "/PATH/TO/sslCredStore.cs" --password clientStorePassword --add example_user --secret examplePwd1!
+----
+
+== Configuring the WildFly Client
+
+Now that we have our key store, trust store, and credential store configured, we need to tell the WildFly client how to
```
Comment (fjuma) [inline _posts/2022-06-15-ejb-over-https.adoc]: s/WildFly client/Elytron client

File: _posts/2022-06-15-ejb-over-https.adoc [hunk 15]
```diff
@@ -0,0 +1,340 @@
...
+access them. From the root of this example, navigate to the directory at `src/main/resources` and open the
+`wildfly-config.xml` configuration file in a text editor. Upon inspecting the file, we note that the credential store
+is imported into the client to access credentials:
```
Comment (fjuma) [inline _posts/2022-06-15-ejb-over-https.adoc]: Just a small terminology thing, instead of "importing", we should reword this to something like "note the credential store configuration. This will be used to access the client's credential when establishing a connection to the server" or something along those lines.

File: _posts/2022-06-15-ejb-over-https.adoc [hunk 16]
```diff
@@ -0,0 +1,340 @@
...
+            <attribute name="location" value="/PATH/TO/sslCredStore.cs"/>
+        </attributes>
+        <protection-parameter-credentials>
+            <clear-password password="clientStorePassword"/>
+        </protection-parameter-credentials>
+    </credential-store>
+</credential-stores>
+----
+
+Similarly, the client also imports the PKCS #12 key store and trust store to access the key pair and server certificate:
```
Comment (fjuma) [inline _posts/2022-06-15-ejb-over-https.adoc]: Same here, instead of "imports", would be good to reword to mention the PKCS #12 key store and trust store configuration.

File: _posts/2022-06-15-ejb-over-https.adoc [hunk 17]
```diff
@@ -0,0 +1,340 @@
...
+        <file name="/PATH/TO/sslClient.keystore"/>
+        <key-store-clear-password password="clientKeySecret"/>
+    </key-store>
+</key-stores>
+----
+
+In all three locations, we will need to replace `/PATH/TO` with the filepath to the various stores.
+
+Looking further into the configuration, we notice the setup for the user authentication under the
+`<authentication-configurations>` tag. The client selects a username, and then retrieves the password from the
```
Comment (fjuma) [inline _posts/2022-06-15-ejb-over-https.adoc]: s/selects a username/uses the configured username

File: _posts/2022-06-15-ejb-over-https.adoc (hunk 2 above)
Comment (jessicarod7) [reply _posts/2022-06-15-ejb-over-https.adoc]: alright, i'll update the directory and related contents in the project as well

Comment (jessicarod7) [reply _posts/2022-06-15-ejb-over-https.adoc]: also renaming files and such to use a `tls` prefix for consistency

File: _posts/2022-06-15-ejb-over-https.adoc [hunk 18]
```diff
@@ -0,0 +1,340 @@
...
+        </credentials>
+        <sasl-mechanism-selector selector="SCRAM-SHA-512-PLUS"/>
+        <providers>
+            <use-service-loader/>
+        </providers>
+    </configuration>
+</authentication-configurations>
+----
+
+Below this is the configuration for the TLS connection, under the `<ssl-contexts>` tag. The client specifies both the
```
Comment (fjuma) [inline _posts/2022-06-15-ejb-over-https.adoc]: s/The client/The configuration

File: _posts/2022-06-15-ejb-over-https.adoc (hunk 8 above)
Comment (jessicarod7) [reply _posts/2022-06-15-ejb-over-https.adoc]: got it, will update

File: _posts/2022-06-15-ejb-over-https.adoc [hunk 19]
```diff
@@ -0,0 +1,340 @@
...
+
+Then, we add the `example_user` identity to the realm, and set a `groups` attribute with the `guest` value :
+
+[source]
+----
+/subsystem=elytron/filesystem-realm=sslFsRealm:add-identity(identity=example_user)
+
+/subsystem=elytron/filesystem-realm=sslFsRealm:set-password(identity=example_user, clear={password=examplePwd1!})
+
+/subsystem=elytron/filesystem-realm=sslFsRealm:add-identity-attribute(identity=example_user, name=groups, value=[guest])
```
Comment (fjuma) [inline _posts/2022-06-15-ejb-over-https.adoc]: Instead of groups, I think you could just use Roles here and then the groups-to-roles decoder won't be needed blow.

File: _posts/2022-06-15-ejb-over-https.adoc [hunk 20]
```diff
@@ -0,0 +1,340 @@
...
+
+Next, we will configure the security domain to use the filesystem realm. We'll use the included `groups-to-roles` role
+decoder to convert the `groups` attribute to a `guest` role:
+
+[source]
+----
+/subsystem=elytron/security-domain=sslFsSD:add(realms=[{realm=sslFsRealm,role-decoder=groups-to-roles}],default-realm=sslFsRealm,permission-mapper=default-permission-mapper)
+----
+
+Afterwards, we add a mapping to the security domain into the `ejb3` subsystem, for setting access based on roles:
```
Comment (fjuma) [inline _posts/2022-06-15-ejb-over-https.adoc]: s/for setting access based on roles/for securing the EJB

**Comment by jessicarod7** [comment]
Blog post can be previewed on my [GitHub Pages](https://cam-rod.github.io/wildfly-elytron/blog/ejb-over-tls/)

**Comment by fjuma** [comment]
Nice! Just a minor suggestion but I think it would be good to add a table of contents to this post to easily see the different sections. This can be done using toc tags as shown here:

https://wildfly-security.github.io/wildfly-elytron/blog/multi-tenancy-support-openid-connect/
https://raw.githubusercontent.com/wildfly-security/wildfly-elytron/develop/_posts/2022-01-21-multi-tenancy-support-openid-connect.adoc

File: _posts/2022-06-15-ejb-over-tls.adoc [hunk 21]
```diff
@@ -0,0 +1,343 @@
...
+date: 2022-06-17
+tags: ejb tls sasl https credential-store keystore truststore
+synopsis: A demonstration on how to configure Elytron for secure connections to EJBs, using mutual TLS, SASL authentication mechanisms, and credential stores.
+author: cam-rod
+---
+
+:toc: macro
+:toc-title:
+
+The *https://github.com/wildfly-security-incubator/elytron-examples/tree/master/ejb-https[ejb-https]* example
```
Comment (fjuma) [inline _posts/2022-06-15-ejb-over-tls.adoc]: s/ejb-https/ejb-mutual-tls

**Comment by fjuma** [comment]
In the TOC, we can see:

* Configuring the Credential Store
  * Configuring the WildFly Client
  * Configuring the Security Domain
  * Enabling SCRAM-SHA-512-PLUS
* Configuring the Remote Connector

To better group things, I wonder if we can update this to something like:

* Configuring the Credential Store
* Configuring the WildFly Client
* Configuring the Server
  * Configuring the Security Domain
  * Enabling SCRAM-SHA-512-PLUS
  * Configuring the Remote Connector

**Review by fjuma — APPROVED** [review]
Thanks @cam-rod!

File: _posts/2022-06-20-ejb-over-tls.adoc [hunk 22]
```diff
@@ -0,0 +1,346 @@
...
+Java *keytool*, we store it in a PKCS #12 keystore:
+
+[source,shell]
+----
+$ keytool -genkeypair -alias client -keyalg RSA -keysize 2048 -validity 365 -keystore /PATH/TO/tlsClient.keystore -dname "CN=client" -storepass clientKeySecret
+----
+
+Note that we have set the common name on the key to `client` (using the option `-dname "CN=client"`). The server will
+try to match the common name on the certificate to the hostname presented during the TLS handshake. If the two do not
+match, the connection will fail.
```
Comment (skyllarr) [inline _posts/2022-06-20-ejb-over-tls.adoc]: @cam-rod Is it true that the connection would fail? In the example, the "client" is not a presented hostname, right?

Comment (jessicarod7) [reply _posts/2022-06-20-ejb-over-tls.adoc]: Since it's mutual TLS, the client presents its cert and hostname. Also had it fail on me when it didn't match in testing.

Comment (skyllarr) [reply _posts/2022-06-20-ejb-over-tls.adoc]: It might be good to rephrase slightly. The server is not checking whether the CN on client's certificate matches the client's hostname, since the client's hostname is not configured. It would be localhost if anything, when running the example on your machine.

As you documented, the server checks whether it trusts the certificate presented by the client, and it does because it is in its truststore. 

When it failed in your testing, maybe you regenerated the client certificate without updating the truststore or the alias was incorrect.

Comment (skyllarr) [reply _posts/2022-06-20-ejb-over-tls.adoc]: clients do often use hostname verification, and then the hostname presented in server's certificate must be the same as the hostname the client is connecting to. Or some other implementation of hostname verification can be used (eg. in Java has an [interface](https://docs.oracle.com/javase/7/docs/api/javax/net/ssl/HostnameVerifier.html) for this also ) But the server does not do this for the client here.

File: _posts/2022-06-20-ejb-over-tls.adoc [hunk 23]
```diff
@@ -0,0 +1,346 @@
...
+        <file name="/PATH/TO/tlsClient.keystore"/>
+        <key-store-clear-password password="clientKeySecret"/>
+    </key-store>
+</key-stores>
+----
+
+In all three locations, we will need to replace `/PATH/TO` with the filepath to the various stores.
+
+Looking further into the configuration, we notice the setup for the user authentication under the
+`<authentication-configurations>` tag. The client uses the configuredusername, and then retrieves the password from the
```
Comment (skyllarr) [inline _posts/2022-06-20-ejb-over-tls.adoc]: Just a minor, the configuredusername is missing space

**Comment by skyllarr** [comment]
@cam-rod Great blog post! I like the explanations and relevant links to other posts. I just added minor comments otherwise LGTM

File: _posts/2022-06-20-ejb-over-tls.adoc (hunk 22 above)
Comment (jessicarod7) [reply _posts/2022-06-20-ejb-over-tls.adoc]: Got it, thanks for the explanation! Moved those sentences to the server certificate generation section.
